shopfloor/actions/stock.py

```diff
@@ -10,6 +13,93 @@ class StockAction(Component):
+                quantity -= sum(move.move_line_ids.mapped("product_qty"))
+            elif move.state in ("done"):
+                quantity -= move.product_qty
+        quantity = float_round(
+            quantity, precision_rounding=origin_move.product_id.uom_id.rounding
+        )
+        return_type = return_picking.picking_type_id
+        return {
+            "product_id": product.id,
+            "product_uom_qty": quantity,
+            "product_uom": product.uom_id.id,
+            "picking_id": return_picking.id,
+            "state": "draft",
+            "date": fields.Datetime.now(),
+            "location_id": return_picking.location_id.id,
+            "location_dest_id": return_picking.location_dest_id.id,
+            "picking_type_id": return_type.id,
+            "warehouse_id": return_type.warehouse_id.id,
+            "origin_returned_move_id": origin_move.id,
+            "procure_method": "make_to_stock",
+        }
+
+    def _create_return_move__link_to_origin(self, return_move, origin_move):
+        move_orig_to_link = origin_move.move_dest_ids.mapped("returned_move_ids")
+        move_orig_to_link |= origin_move
+        origin_move_dest = origin_move.move_dest_ids.filtered(
+            lambda m: m.state not in ("cancel")
+        )
+        move_orig_to_link |= origin_move_dest.move_orig_ids.filtered(
+            lambda m: m.state not in ("cancel")
+        )
+        move_dest_to_link = origin_move.move_orig_ids.mapped("returned_move_ids")
+        move_dest_orig = origin_move.returned_move_ids.move_orig_ids.filtered(
+            lambda m: m.state not in ("cancel")
+        )
+        move_dest_to_link |= move_dest_orig.move_dest_ids.filtered(
+            lambda m: m.state not in ("cancel")
+        )
+        write_vals = {
+            "move_orig_ids": [(4, m.id) for m in move_orig_to_link],
+            "move_dest_ids": [(4, m.id) for m in move_dest_to_link],
+        }
+        return_move.write(write_vals)
+
+    def create_return_move(self, return_picking, origin_move):
+        """Creates a return move for a given return picking / move"""
+        # Logic has been copied from
+        # odoo_src/addons/stock/wizard/stock_picking_return.py
+        return_move_vals = self._create_return_move__get_vals(
+            return_picking, origin_move
+        )
+        return_move = origin_move.copy(return_move_vals)
+        self._create_return_move__link_to_origin(return_move, origin_move)
+        return return_move
+
+    def _create_return_picking__get_vals(self, picking, return_types):
+        return_type = (
+            fields.first(return_types) or picking.picking_type_id.return_picking_type_id
+        )
+        return {
+            "move_lines": [],
+            "picking_type_id": return_type.id,
+            "state": "draft",
+            # This isn't std but helps in `scan_line`.
+            # f"Return of  {outgoing_picking.name}"
+            "origin": picking.name,
```

Comment:
I don't like this. Especially that you reach here after taking the first matching picking which could be wrong as you don't know yet which product.
We should not loose the origin and probably reuse it to collect back the complete set of valid pickings and not just the first one